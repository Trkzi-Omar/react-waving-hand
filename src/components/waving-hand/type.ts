import type { HTMLAttributes } from 'react'

export enum FSP {
  I = '👋🏻',
  II = '👋🏼',
  III = '👋',
  IV = '👋🏽',
  V = '👋🏾',
  VI = '👋🏿'
}
export type Phototype = keyof typeof FSP

export type NaturalNumber = number & { __naturalNumber: true }
export type NumericPhototype = 1 | 2 | 3 | 4 | 5 | 6

interface Props {
  size?: string
  type?: NumericPhototype
  motion?: boolean
  loop?: number
}
export type MixedProps = Props & HTMLAttributes<HTMLSpanElement>
