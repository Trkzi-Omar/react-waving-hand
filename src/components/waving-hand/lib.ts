import type { NaturalNumber, NumericPhototype, Phototype } from './type'

const TOKEN_MAP: Readonly<Record<NumericPhototype, Phototype>> = {
  1: 'I',
  2: 'II',
  3: 'III',
  4: 'IV',
  5: 'V',
  6: 'VI'
}

export const positiveNumberSchema = (x: number) => {
  if (x >= 0) return x as NaturalNumber
  throw new Error(
    '[loop] property needs to be a natural number which means it needs to be equal to or greater than 0.'
  )
}

export const romanize = (num: NumericPhototype) => TOKEN_MAP[num]
