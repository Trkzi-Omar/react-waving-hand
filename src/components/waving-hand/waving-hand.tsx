import './style.css'

import { memo, useMemo, CSSProperties } from 'react'

import { positiveNumberSchema, romanize } from './lib'
import { FSP, MixedProps } from './type'

const WavingHand = ({
  size = '1rem',
  type = 3,
  motion = false,
  loop = 0,
  style,
  className,
  ...restAttributes
}: MixedProps) => {
  const computedStyle = useMemo<CSSProperties>(() => {
    const animationIterationCount = loop !== 0 && motion ? positiveNumberSchema(loop) : 'infinite'

    return {
      fontSize: size,
      ...(motion && { animationIterationCount }),
      ...style
    }
  }, [size, motion, loop, style])

  const computedClassName = useMemo<string>(
    () => [motion ? 'motion' : '', className].filter(Boolean).join(' '),
    [motion, className]
  )

  const content = useMemo<FSP>(() => FSP[romanize(type)], [type])

  return (
    <span style={computedStyle} className={computedClassName} {...restAttributes}>
      {content}
    </span>
  )
}

const MemoizedWavingHand = memo(WavingHand)
export { MemoizedWavingHand as WavingHand }
