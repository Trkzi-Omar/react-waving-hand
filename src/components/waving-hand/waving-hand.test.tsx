import * as renderer from 'react-test-renderer'

import { WavingHand } from './waving-hand'

it('turn on motion', () => {
  const tree = renderer
    .create(<WavingHand loop={2} motion={true} size='2.5rem' type={2} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('turn off motion', () => {
  const tree = renderer.create(<WavingHand />).toJSON()
  expect(tree).toMatchSnapshot()
})
