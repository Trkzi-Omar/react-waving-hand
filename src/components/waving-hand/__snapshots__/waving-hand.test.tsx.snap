// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props testing turn off motion 1`] = `
<span
  className=""
  style={
    {
      "fontSize": "1rem",
    }
  }
>
  👋
</span>
`;

exports[`props testing turn on motion 1`] = `
<span
  className="motion"
  style={
    {
      "animationIterationCount": 2,
      "fontSize": "2.5rem",
    }
  }
>
  👋🏼
</span>
`;
